pkg/s3select/input.go

```diff
@@ -0,0 +1,338 @@
+	}
+	statXML := &stats{
+		BytesScanned:   reader.stats.BytesScanned,
+		BytesProcessed: reader.stats.BytesProcessed,
+		BytesReturned:  reader.stats.BytesReturned,
+	}
+	out, err := xml.Marshal(statXML)
+	if err != nil {
+		return "", err
+	}
+	return xml.Header + string(out), nil
+}
+
+// createProgressXML is the function which does the marshaling from the progress structs into XML so that the progress and stat message can be sent
+func (reader *Input) createProgressXML() (string, error) {
+	if reader.options.HasHeader {
+		reader.stats.BytesProcessed += processSize(reader.header)
+	}
+	if !reader.options.CompressedGzip {
+		reader.stats.BytesScanned = reader.stats.BytesProcessed
+	}
+	progressXML := &progress{
+		BytesScanned:   reader.stats.BytesScanned,
+		BytesProcessed: reader.stats.BytesProcessed,
+		BytesReturned:  reader.stats.BytesReturned,
+	}
+	out, err := xml.Marshal(progressXML)
+	if err != nil {
+		return "", err
+	}
+	return xml.Header + string(out), nil
+}
+
+// Header returns the header of the reader. Either the first row if a header
+// set in the options, or c#, where # is the column number, starting with 0.
+func (reader *Input) Header() []string {
+	return reader.header
+}
+
+// Row is a Struct for keeping track of key aspects of a row.
+type Row struct {
+	record string
+	err    error
+}
+
+// Execute is the function where all the blocking occurs, It writes to the HTTP
+// response writer in a streaming fashion so that the client can actively use
+// the results before the query is finally finished executing. The
+func (reader *Input) Execute(writer http.ResponseWriter) error {
+	myRow := make(chan *Row)
+	progressTicker := createTicker(60 * time.Second)
+	continuationTicker := createTicker(180 * time.Second)
+	var curBuf = &bytes.Buffer{}
+	go reader.runSelectParser(convertMySQL(reader.options.Expression), myRow)
+	for {
+		select {
+		case row, ok := <-myRow:
+			if ok && len(row.record) > 0 {
+				message := reader.writeRecordMessage(row.record, curBuf)
+				message.WriteTo(writer)
+				curBuf.Reset()
+				reader.stats.BytesReturned += int64(len(row.record))
+				progressTicker.resetTicker()
+				continuationTicker.resetTicker()
+			} else if ok && row.err != nil {
+				errorMessage := reader.writeErrorMessage(row.err, curBuf)
+				errorMessage.WriteTo(writer)
+				curBuf.Reset()
+				close(myRow)
+				return nil
+			} else if !ok {
+				statPayload, err := reader.createStatXML()
+				if err != nil {
+					return err
+				}
+				statMessage := reader.writeStatMessage(statPayload, curBuf)
+				statMessage.WriteTo(writer)
+				curBuf.Reset()
+				message := reader.writeEndMessage(curBuf)
+				message.WriteTo(writer)
+				return nil
+			}
+
+		case <-progressTicker.ticker.C:
+			progressPayload, err := reader.createProgressXML()
+			if err != nil {
+				return err
+			}
+			progressMessage := reader.writeProgressMessage(progressPayload, curBuf)
+			progressMessage.WriteTo(writer)
```

Comment:
Does this mean that progress and result messages are sent interleaved to the client?

Reply:
Yes @donatello since its a protocol each payload and header signify what they are.